File: 6-todo-app.py
import sqlite3

# เชื่อมต่อ database
conn = sqlite3.connect('database.db')
cursor = conn.cursor()

# ==========================================
# Todo App Functions
# ==========================================

class TodoApp:
    def __init__(self, cursor, conn):
        self.cursor = cursor
        self.conn = conn
    
    # เพิ่ม todo ใหม่
    def add_todo(self, task):
        self.cursor.execute('INSERT INTO todos (task) VALUES (?)', (task,))
        self.conn.commit()
        todo_id = self.cursor.lastrowid
        print(f'✅ Added: "{task}" (ID: {todo_id})')
    
    # แสดง todos ทั้งหมด
    def show_all(self):
        self.cursor.execute('SELECT * FROM todos')
        todos = self.cursor.fetchall()
        print('\n📋 All Todos:')
        print(f"{'ID':>4} {'Task':<30} {'Done':>6} {'Created At':<20}")
        print('-' * 65)
        for todo in todos:
            print(f"{todo[0]:>4} {todo[1]:<30} {todo[2]:>6} {todo[3]:<20}")
    
    # แสดง todos ที่ยังไม่เสร็จ
    def show_pending(self):
        self.cursor.execute('SELECT * FROM todos WHERE done = 0')
        todos = self.cursor.fetchall()
        print('\n⏳ Pending Todos:')
        print(f"{'ID':>4} {'Task':<30}")
        print('-' * 40)
        for todo in todos:
            print(f"{todo[0]:>4} {todo[1]:<30}")
    
    # แสดง todos ที่เสร็จแล้ว
    def show_completed(self):
        self.cursor.execute('SELECT * FROM todos WHERE done = 1')
        todos = self.cursor.fetchall()
        print('\n✅ Completed Todos:')
        print(f"{'ID':>4} {'Task':<30}")
        print('-' * 40)
        for todo in todos:
            print(f"{todo[0]:>4} {todo[1]:<30}")
    
    # แสดง todos เรียงตามวันที่สร้าง (ใหม่สุดก่อน)
    def show_by_date(self):
        self.cursor.execute('SELECT * FROM todos ORDER BY created_at DESC')
        todos = self.cursor.fetchall()
        print('\n🗓️ Todos by date (newest first):')
        print(f"{'ID':>4} {'Task':<30} {'Created At':<20}")
        print('-' * 60)
        for todo in todos:
            print(f"{todo[0]:>4} {todo[1]:<30} {todo[3]:<20}")
    
    # ทำเครื่องหมายว่าเสร็จ
    def mark_as_done(self, todo_id):
        self.cursor.execute('UPDATE todos SET done = 1 WHERE id = ?', (todo_id,))
        self.conn.commit()
        if self.cursor.rowcount > 0:
            print(f'✅ Marked todo #{todo_id} as done')
        else:
            print(f'❌ Todo #{todo_id} not found')
    
    # ค้นหา todos จาก keyword
    def search_todos(self, keyword):
        self.cursor.execute('SELECT * FROM todos WHERE task LIKE ?', (f'%{keyword}%',))
        todos = self.cursor.fetchall()
        print(f'\n🔎 Search results for "{keyword}":')
        print(f"{'ID':>4} {'Task':<30} {'Done':>6}")
        print('-' * 45)
        for todo in todos:
            print(f"{todo[0]:>4} {todo[1]:<30} {todo[2]:>6}")
    
    # แก้ไขข้อความของ todo
    def update_task(self, todo_id, new_task):
        self.cursor.execute('UPDATE todos SET task = ? WHERE id = ?', (new_task, todo_id))
        self.conn.commit()
        if self.cursor.rowcount > 0:
            print(f'✏️ Updated todo #{todo_id}')
        else:
            print(f'❌ Todo #{todo_id} not found')
    
    # ลบ todos ที่เสร็จทั้งหมด
    def clear_completed(self):
        self.cursor.execute('DELETE FROM todos WHERE done = 1')
        self.conn.commit()
        print(f'🧹 Cleared {self.cursor.rowcount} completed todos')
    
    # ลบ todo
    def delete_todo(self, todo_id):
        self.cursor.execute('DELETE FROM todos WHERE id = ?', (todo_id,))
        self.conn.commit()
        if self.cursor.rowcount > 0:
            print(f'🗑️ Deleted todo #{todo_id}')
        else:
            print(f'❌ Todo #{todo_id} not found')
    
    # แสดงสถิติ
    def show_stats(self):
        self.cursor.execute('SELECT COUNT(*) FROM todos')
        total = self.cursor.fetchone()[0]
        
        self.cursor.execute('SELECT COUNT(*) FROM todos WHERE done = 1')
        completed = self.cursor.fetchone()[0]
        
        self.cursor.execute('SELECT COUNT(*) FROM todos WHERE done = 0')
        pending = self.cursor.fetchone()[0]
        
        print('\n📊 Statistics:')
        print(f'  Total: {total}')
        print(f'  ✅ Completed: {completed}')
        print(f'  ⏳ Pending: {pending}')

# ==========================================
# ทดสอบใช้งาน
# ==========================================

app = TodoApp(cursor, conn)

print('🎮 Todo App Demo')
print('=' * 50)

# แสดงทั้งหมด
app.show_all()

# แสดงสถิติ
app.show_stats()

# แสดงที่ยังไม่เสร็จ
app.show_pending()

# ทำเครื่องหมายบางรายการว่าเสร็จ
app.mark_as_done(2)
app.mark_as_done(3)

# แสดงที่เสร็จแล้ว
app.show_completed()

# แสดงสถิติใหม่
app.show_stats()

# ทดสอบ Challenge Features
print('\n' + '=' * 50)
print('🎯 Testing Challenge Features')
print('=' * 50)

# Challenge 1: ค้นหา
app.search_todos('การบ้าน')

# Challenge 2: แก้ไข task
app.update_task(1, 'ซื้อของที่ตลาดและร้านขายยา - อัพเดท')

# Challenge 3: ลบที่เสร็จหมด (ลองเมื่อมีที่เสร็จแล้ว)
# app.clear_completed()

# Challenge 4: เรียงลำดับตามวันที่
app.show_by_date()

# ปิดการเชื่อมต่อ
print()
conn.close()

File: 6-todo-app.js
const Database = require('better-sqlite3');
const db = new Database('database.db');

// ==========================================
// Todo App Functions
// ==========================================

class TodoApp {
  // เพิ่ม todo ใหม่
  addTodo(task) {
    const insert = db.prepare('INSERT INTO todos (task) VALUES (?)');
    const result = insert.run(task);
    console.log(`✅ Added: "${task}" (ID: ${result.lastInsertRowid})`);
  }

  // แสดง todos ทั้งหมด
  showAll() {
    const todos = db.prepare('SELECT * FROM todos').all();
    console.log('\n📋 All Todos:');
    console.table(todos);
  }

  // แสดง todos ที่ยังไม่เสร็จ
  showPending() {
    const todos = db.prepare('SELECT * FROM todos WHERE done = 0').all();
    console.log('\n⏳ Pending Todos:');
    console.table(todos);
  }

  // แสดง todos ที่เสร็จแล้ว
  showCompleted() {
    const todos = db.prepare('SELECT * FROM todos WHERE done = 1').all();
    console.log('\n✅ Completed Todos:');
    console.table(todos);
  }

  // แสดง todos เรียงตามวันที่สร้าง (ใหม่สุดก่อน)
  showByDate() {
    const todos = db.prepare('SELECT * FROM todos ORDER BY created_at DESC').all();
    console.log('\n🗓️ Todos by date (newest first):');
    console.table(todos);
  }

  // ทำเครื่องหมายว่าเสร็จ
  markAsDone(id) {
    const update = db.prepare('UPDATE todos SET done = 1 WHERE id = ?');
    const result = update.run(id);
    if (result.changes > 0) {
      console.log(`✅ Marked todo #${id} as done`);
    } else {
      console.log(`❌ Todo #${id} not found`);
    }
  }

  // ค้นหา todos จาก keyword
  searchTodos(keyword) {
    const search = db.prepare('SELECT * FROM todos WHERE task LIKE ?');
    const todos = search.all(`%${keyword}%`);
    console.log(`\n🔎 Search results for "${keyword}":`);
    console.table(todos);
  }

  // แก้ไขข้อความของ todo
  updateTask(id, newTask) {
    const update = db.prepare('UPDATE todos SET task = ? WHERE id = ?');
    const result = update.run(newTask, id);
    if (result.changes > 0) {
      console.log(`✏️ Updated todo #${id}`);
    } else {
      console.log(`❌ Todo #${id} not found`);
    }
  }

  // ลบ todos ที่เสร็จทั้งหมด
  clearCompleted() {
    const del = db.prepare('DELETE FROM todos WHERE done = 1');
    const result = del.run();
    console.log(`🧹 Cleared ${result.changes} completed todos`);
  }

  // ลบ todo
  deleteTodo(id) {
    const del = db.prepare('DELETE FROM todos WHERE id = ?');
    const result = del.run(id);
    if (result.changes > 0) {
      console.log(`🗑️ Deleted todo #${id}`);
    } else {
      console.log(`❌ Todo #${id} not found`);
    }
  }

  // แสดงสถิติ
  showStats() {
    const total = db.prepare('SELECT COUNT(*) as count FROM todos').get();
    const completed = db.prepare('SELECT COUNT(*) as count FROM todos WHERE done = 1').get();
    const pending = db.prepare('SELECT COUNT(*) as count FROM todos WHERE done = 0').get();

    console.log('\n📊 Statistics:');
    console.log(`  Total: ${total.count}`);
    console.log(`  ✅ Completed: ${completed.count}`);
    console.log(`  ⏳ Pending: ${pending.count}`);
  }
}

// ==========================================
// ทดสอบใช้งาน
// ==========================================

const app = new TodoApp();

console.log('🎮 Todo App Demo');
console.log('='.repeat(50));

// แสดงทั้งหมด
app.showAll();

// แสดงสถิติ
app.showStats();

// แสดงที่ยังไม่เสร็จ
app.showPending();

// ทำเครื่องหมายบางรายการว่าเสร็จ
app.markAsDone(2);
app.markAsDone(3);

// แสดงที่เสร็จแล้ว
app.showCompleted();

// แสดงสถิติใหม่
app.showStats();

// ปิดการเชื่อมต่อ
console.log('');
db.close();
